### PR TITLE
no-changed-when: improve testing and documentation

### PR DESCRIPTION
**Pull request by konstruktoid**
- adds tests to the `no-changed-when` rule
- improve documentation

Signed-off-by: Thomas Sjögren <konstruktoid@users.noreply.github.com>
Fixes: #1681

### CODE REVIEW

File: src/ansiblelint/rules/CommandHasChangesCheckRule.py [hunk 1]
```diff
@@ -31,12 +32,30 @@
...
+done (using ``creates`` or ``removes`` argument) or only do it if another
+check has a particular result (``when``).
+
+For example, this task registers the ``shell`` output and uses the return code
+to define when the task has changed.
+
+.. code:: yaml
+
+    - name: handle shell output with return code
+      ansible.builtin.shell: cat {{ myfile|quote }}
```
Comment (ssbarnea) [inline src/ansiblelint/rules/CommandHasChangesCheckRule.py:47]: This needs quotes, lets avoid examples that would not pass linting.

File: src/ansiblelint/rules/CommandHasChangesCheckRule.py [hunk 2]
```diff
@@ -31,12 +32,30 @@
...
+      register: myoutput
+      changed_when: myoutput.rc != 0
+
+The following example will trigger the rule since the task does not
+handle the output of the ``command``.
+
+.. code:: yaml
+
+    - name: does not handle any output or return codes
+      ansible.builtin.command: cat {{ myfile|quote }}
```
Comment (ssbarnea) [inline src/ansiblelint/rules/CommandHasChangesCheckRule.py:57]: DITTO, add quotes to this due to use of jinja braces.

**Comment by MarkusTeufelberger** [comment]
The existing `when` case is wrong by the way (the presence of `when: true` does not influence the idempotence of a task result), but that likely needs a separate PR to be fixed. It could need a test case though, if you're already writing tests.

File: src/ansiblelint/rules/CommandHasChangesCheckRule.py (hunk 1 above)
Comment (konstruktoid) [reply src/ansiblelint/rules/CommandHasChangesCheckRule.py:47]: Except that it does.

```sh
$ ansible-lint --version
ansible-lint 5.1.3.dev22 using ansible 2.11.4
$ cat /tmp/issue1681.yml
---
- hosts: all
  tasks:
    - name: command without quotes
      ansible.builtin.command: cat {{ myfile|quote }}
      register: myoutput
      changed_when: myoutput.rc != 0

    - name: command should fail
      ansible.builtin.command: cat {{ myfile|quote }}
...
$ ansible-lint /tmp/issue1681.yml
Loading custom .yamllint config file, this extends our internal yamllint config.
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: /tmp/issue1681.yml
WARNING  Listing 1 violation(s) that are fatal
no-changed-when: Commands should not change things if nothing needs doing
../../../../tmp/issue1681.yml:9 Task/Handler: command should fail

You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
  - no-changed-when  # Commands should not change things if nothing needs doing

Finished with 1 failure(s), 0 warning(s) on 1 files.
$
```

**Comment by konstruktoid** [comment]
> 
> 
> The existing `when` case is wrong by the way (the presence of `when: true` does not influence the idempotence of a task result), but that likely needs a separate PR to be fixed. It could need a test case though, if you're already writing tests.

Yeah, but I believe that is out-of-scope for this PR.

**Pull request merged**